checker/src/org/checkerframework/checker/signedness/SignednessVisitor.java

```diff
@@ -23,6 +22,124 @@ public SignednessVisitor(BaseTypeChecker checker) {
+     * @param numBitsLit the LiteralTree whose value is numBits
+     * @param maskLit the LiteralTree whose value is mask
+     * @return true iff the numBits most significant bits of mask are 0 for AND, and 1 for OR
+     */
+    private boolean maskIgnoresMSB(Kind maskKind, LiteralTree numBitsLit, LiteralTree maskLit) {
+        long numBits = getLong(numBitsLit.getValue());
+        long mask = getLong(maskLit.getValue());
+
+        // If maskLit was an int, then shift mask so that the numBits most significant bits are in the right position
+        if (maskLit.getKind() != Kind.LONG_LITERAL) {
+            mask <<= 32;
+        }
+
+        // Shift the numBits most significant bits to become the numBits least significant bits, zeroing out the rest
+        mask >>>= (64 - numBits);
+        if (maskKind == Kind.AND) {
+            // Check that the numBits most significant bits of the mask were 0
+            return mask == 0;
+        } else if (maskKind == Kind.OR) {
+            // Check that the numBits most significant bits of the mask were 1
+            return mask == (1 << numBits) - 1;
+        } else {
+            return false;
+        }
+    }
+
+    /**
+     * Determines if a right shift operation, {@code >>} or {@code >>>}, is masked with a masking
+     * operation of the form {@code shiftOp & maskLit} or {@code shiftOp | maskLit} such that the
+     * mask renders the shift signedness irrelevent by destroying the bits introduced by the shift.
+     * For example, the following pairs of right shifts on {@code byte b} both produce the same
+     * results under any input, because of their masks:
+     *
+     * <p>{@code (b >> 4) & 0x0F == (b >>> 4) & 0x0F;} {@code (b >> 4) | 0xF0 == (b >>> 4) | 0xF0;}
+     *
+     * @param shiftOp a right shift operation: {@code >>} or {@code >>>}
+     * @return true iff the right shift is masked such that a signed or unsigned right shift has the
+     *     same effect
+     */
+    private boolean isMaskedShift(BinaryTree shiftOp) {
+        // parent is the operation or statement that immediately contains shiftOp
+        Tree parent;
+        // topChild is the top node in the chain of nodes from shiftOp to parent
+        Tree topChild;
+        {
+            TreePath parentPath = visitorState.getPath().getParentPath();
+            parent = parentPath.getLeaf();
+            topChild = parent;
+            // Strip away all parentheses from the shift operation
+            while (parent.getKind() == Kind.PARENTHESIZED) {
+                parentPath = parentPath.getParentPath();
+                topChild = parent;
+                parent = parentPath.getLeaf();
+            }
+        }
+
+        if (!isMask(parent)) {
+            return false;
+        }
+
+        BinaryTree maskOp = (BinaryTree) parent;
+        ExpressionTree shiftExpr = shiftOp.getRightOperand();
+
+        // Determine which child of maskOp leads to shiftOp. The other one is the mask expression
```

Comment:
Throughout, whenever you have a full sentence, terminate it with punctuation (here, a period).  Otherwise, it's too confusing to read -- I was taken aback and had to slow down and reread to realize there was nothing missing here.